src/target/xtensa/xtensa.c

```diff
@@ -652,31 +637,81 @@ static int xtensa_write_dirty_registers(struct target *target)
-		xtensa->core_config->queue_write_dirty_user_regs(target);
+	if (delay_cpenable) {
+		regval = xtensa_reg_get(target, XT_REG_IDX_CPENABLE);
+		LOG_TARGET_DEBUG(target, "Writing back reg cpenable (224) val %08" PRIX32, regval);
+		xtensa_queue_dbg_reg_write(xtensa, NARADR_DDR, regval);
+		xtensa_queue_exec_ins(xtensa, XT_INS_RSR(xtensa, XT_SR_DDR, XT_REG_A3));
+		xtensa_queue_exec_ins(xtensa, XT_INS_WSR(xtensa, 
+			xtensa_regs[XT_REG_IDX_CPENABLE].reg_num, 
+			XT_REG_A3));
+		reg_list[XT_REG_IDX_CPENABLE].dirty = false;
+	}
+
+	preserve_a3 = (xtensa->core_config->windowed);
+	if (preserve_a3) {
+		/* Save (windowed) A3 for scratch use */
+		xtensa_queue_exec_ins(xtensa, XT_INS_WSR(xtensa, XT_SR_DDR, XT_REG_A3));
+		xtensa_queue_dbg_reg_read(xtensa, NARADR_DDR, a3_buf);
+		res = jtag_execute_queue();
+		xtensa_core_status_check(target);
+		a3 = buf_get_u32(a3_buf, 0, 32);
+	}
 
 	if (xtensa->core_config->windowed) {
-		/*Grab the windowbase, we need it. */
+		res = xtensa_window_state_save(target, &woe);
+		if (res != ERROR_OK)
+			return res;
+		/* Grab the windowbase, we need it. */
 		windowbase = xtensa_reg_get(target, XT_REG_IDX_WINDOWBASE);
-		/*Check if there are problems with both the ARx as well as the corresponding Rx
-		 * registers set and dirty. */
-		/*Warn the user if this happens, not much else we can do... */
+		/* Check if there are mismatches between the ARx and corresponding Ax registers.
+		 * When the user sets a register on a windowed config, xt-gdb may set the ARx
+		 * register directly.  Thus we take ARx as priority over Ax if both are dirty
+		 * and it's unclear if the user set one over the other explicitly.
+		 */
 		for (unsigned int i = XT_REG_IDX_A0; i <= XT_REG_IDX_A15; i++) {
-			unsigned int j = xtensa_windowbase_offset_to_canonical(i, windowbase);
+			unsigned int j = xtensa_windowbase_offset_to_canonical(xtensa, i, windowbase);
 			if (reg_list[i].dirty && reg_list[j].dirty) {
-				if (memcmp(reg_list[i].value, reg_list[j].value,
-						sizeof(xtensa_reg_val_t)) != 0)
-					LOG_WARNING(
-						"Warning: Both A%d as well as the physical register it points to (AR%d) are dirty and differs in value. Results are undefined!",
-						i - XT_REG_IDX_A0,
-						j - XT_REG_IDX_AR0);
+				if (memcmp(reg_list[i].value, reg_list[j].value, sizeof(xtensa_reg_val_t)) != 0) {
+					bool show_warning = true;
+					if (i == XT_REG_IDX_A3) {
+						if (xtensa->scratch_ars[XT_AR_SCRATCH_A3].intval &&
+							!xtensa->scratch_ars[XT_AR_SCRATCH_AR3].intval) {
+							LOG_DEBUG("AR conflict: a3 -> ar%d", j - XT_REG_IDX_AR0);
+							memcpy(reg_list[j].value, reg_list[i].value, sizeof(xtensa_reg_val_t));
+						} else {
+							LOG_DEBUG("AR conflict: ar%d -> a3", j - XT_REG_IDX_AR0);
+							memcpy(reg_list[i].value, reg_list[j].value, sizeof(xtensa_reg_val_t));
+						}
+						show_warning = (xtensa->scratch_ars[XT_AR_SCRATCH_A3].intval &&
+							xtensa->scratch_ars[XT_AR_SCRATCH_AR3].intval);
```

Comment:
I'd wrapup this into function, smth like
```suggestion
                                                        show_warning = xtensa_scratch_reqs_fixup(xtensa, reg_list, i, j, XT_AR_SCRATCH_A3, XT_AR_SCRATCH_AR3);
```
```c
void xtensa_scratch_reqs_fixup(struct xtensa *xtensa, struct reg *reg_list, int i, int j, int a_idx, int ar_idx)
{
  if (xtensa->scratch_ars[a_idx].intval &&
	!xtensa->scratch_ars[ar_idx].intval) {
    LOG_DEBUG("AR conflict: a3 -> ar%d", j - XT_REG_IDX_AR0);
    memcpy(reg_list[j].value, reg_list[i].value, sizeof(xtensa_reg_val_t));
  } else {
    LOG_DEBUG("AR conflict: ar%d -> a3", j - XT_REG_IDX_AR0);
    memcpy(reg_list[i].value, reg_list[j].value, sizeof(xtensa_reg_val_t));
  }
  return (xtensa->scratch_ars[a_idx].intval &&
		xtensa->scratch_ars[ar_idx].intval);
}
```
and re-use it below

Reply:
Yes, good idea.  I'll rework this.  Thanks!